### PR TITLE
bridge: packet corruption and reordering

### PR DESCRIPTION
**Pull request by heyitsanthony**
With bonus bridge connection code refactor.

### CODE REVIEW

File: tools/local-tester/bridge/dispatch.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+func newDispatcherPool() dispatcher {
+	d := &dispatcherPool{}
+	go d.writeLoop()
+	return d
+}
+
+func (d *dispatcherPool) writeLoop() {
+	for {
+		time.Sleep(100 * time.Millisecond)
```
Comment (gyuho) [inline tools/local-tester/bridge/dispatch.go]: did you mean to use `dispatchPoolDelay` here? `time.Sleep(dispatchPoolDelay)`?

File: tools/local-tester/bridge/bridge.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2016 The etcd Authors
+// Dispatchright 2016 The etcd Authors
```
Comment (xiang90) [inline tools/local-tester/bridge/bridge.go]: what is dispatchright? auto replacing is dangerous.

Comment (heyitsanthony) [inline tools/local-tester/bridge/bridge.go]: oops...

**Comment by gyuho** [comment]
lgtm. defer to @xiang90 

thanks.

**Comment by heyitsanthony** [comment]
Unsurprisingly, lots of panics on peer connection corruption with the worst being snapshot corruption permanently downing the node. Plain reordering without corruption seems stable.

**Comment by xiang90** [comment]
@heyitsanthony I do not think we have ANY mechanism to deal with packet corruption right now.

**Comment by xiang90** [comment]
We have some mechanisms to handle disk corruption though.

**Comment by heyitsanthony** [comment]
@xiang90 yeah, that's why it's not a surprise...

File: tools/local-tester/bridge/dispatch.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+// dispatchPacketBytes is how many bytes to send until choosing a new connection
+var dispatchPacketBytes = 32
+
+type dispatcher interface {
+	// Copy works like io.Copy using buffers provided by fetchFunc
+	Copy(io.Writer, fetchFunc) error
+}
+
+type fetchFunc func() ([]byte, error)
```
Comment (xiang90) [inline tools/local-tester/bridge/dispatch.go]: why not io.reader?

File: tools/local-tester/bridge/dispatch.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+import (
+	"io"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+// dispatchPoolDelay is the time to wait before flushing all buffered packets
+var dispatchPoolDelay = 100 * time.Millisecond
```
Comment (xiang90) [inline tools/local-tester/bridge/dispatch.go]:
```go
var (
    ....
)
```

File: tools/local-tester/bridge/dispatch.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+type dispatcher interface {
+	// Copy works like io.Copy using buffers provided by fetchFunc
+	Copy(io.Writer, fetchFunc) error
+}
+
+type fetchFunc func() ([]byte, error)
+
+type dispatcherPool struct {
+	q  []dispatchPacket
+	mu sync.Mutex
```
Comment (xiang90) [inline tools/local-tester/bridge/dispatch.go]: move the mu above the variable it protects?

File: tools/local-tester/bridge/dispatch.go (hunk 3 above)
Comment (heyitsanthony) [inline tools/local-tester/bridge/dispatch.go]: I wanted to leave buffer sizing to the fetchFunc.

File: tools/local-tester/bridge/dispatch.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+func (d *dispatcherPool) flush() {
+	d.mu.Lock()
+	pkts := d.q
+	d.q = nil
+	d.mu.Unlock()
+	if len(pkts) == 0 {
+		return
+	}
+
+	// sort by sockets
```
Comment (xiang90) [inline tools/local-tester/bridge/dispatch.go]: // sort by sockets, preserve the packets ordering inside one socket.

File: tools/local-tester/bridge/dispatch.go (hunk 3 above)
Comment (xiang90) [inline tools/local-tester/bridge/dispatch.go]: ok

**Comment by xiang90** [comment]
The dispatcher code is not efficient due to the extensive copying around. But I do not think it is a big issue either. LGTM.

**Pull request merged**